C/STrace/EtwLogger.cpp

```diff
@@ -0,0 +1,460 @@
+	EventDataDescCreate(&m_providerMetadataDesc, providerMetadata, providerMetadata->TotalLength);
+	m_providerMetadataDesc.Type = EVENT_DATA_DESCRIPTOR_TYPE_PROVIDER_METADATA;  // Descriptor contains provider metadata.
+
+	return STATUS_SUCCESS;
+}
+
+void EtwProvider::Destruct()
+{
+	for (auto i = 0; i < m_events.len(); i++)
+	{
+		m_events[i].Destruct();
+	}
+	m_events.Destruct();
+
+	if (m_regHandle != 0)
+	{
+		EtwUnregister(m_regHandle);
+	}
+
+	if (m_providerMetadataDesc.Ptr != NULL)
+	{
+		ExFreePool((PVOID)m_providerMetadataDesc.Ptr);
+	}
+}
+
+NTSTATUS EtwProvider::AddEvent(const char* eventName, int numberOfFields, va_list fields)
+{
+	auto status = STATUS_SUCCESS;
+
+	if (FindEvent(eventName) == NULL)
+	{
+		EtwProviderEvent event;
+		status = event.Initialize(eventName, numberOfFields, fields);
+		if (status != STATUS_SUCCESS)
+		{
+			return status;
+		}
+		m_events.push_back(move(event));
+	}
+
+	return status;
+}
+
+NTSTATUS EtwProvider::WriteEvent(const char* eventName, uint8_t eventLevel, uint64_t keyword, int numberOfFields, va_list fields)
+{
+	// Find the event to use.
+	const auto event = FindEvent(eventName);
+	if (event == NULL)
+	{
+		return STATUS_UNSUCCESSFUL;
+	}
+
+	// Allocate space for the data descriptors, including two additional slots
+	// for provider and event metadata.
+	const auto numberOfDescriptors = numberOfFields + 2;
+	const auto allocSize = numberOfDescriptors * sizeof(EVENT_DATA_DESCRIPTOR);
+	const auto dataDescriptors = (PEVENT_DATA_DESCRIPTOR)ExAllocatePoolWithTag(NonPagedPoolNx, allocSize, 'wteP');
+	if (dataDescriptors == NULL)
+	{
+		return STATUS_UNSUCCESSFUL;
+	}
+	memset(dataDescriptors, 0, allocSize);
+
+	// Set the provider and event metadata descriptors.
+	dataDescriptors[0] = m_providerMetadataDesc;
+	dataDescriptors[1] = event->MetadataDescriptor();
+
+	// Create a descriptor for each individual field.
+	va_list args;
+	va_copy(args, fields);
+	for (auto i = 0; i < numberOfFields; i++)
+	{
+		va_arg(args, const char*);
+		const auto fieldType = va_arg(args, int);
+		auto fieldValue = va_arg(args, size_t);
+
+		dataDescriptors[i + 2] = CreateTraceProperty(
+			fieldType,
+			fieldType != EtwFieldAnsiString ? &fieldValue : (void*)fieldValue);
+		if (dataDescriptors[i + 2].Ptr == NULL)
+		{
+			return STATUS_UNSUCCESSFUL;
+		}
+	}
+	va_end(args);
+
+	// Create the event descriptor.
+	EVENT_DESCRIPTOR eventDescriptor;
+	memset(&eventDescriptor, 0, sizeof(EVENT_DESCRIPTOR));
+	eventDescriptor.Channel = 11;  // All "manifest-free" events should go to channel 11 by default
+	eventDescriptor.Keyword = keyword;
+	eventDescriptor.Level = eventLevel;
+
+	// Write the event.
+	return EtwWrite(m_regHandle, &eventDescriptor, NULL, numberOfDescriptors, dataDescriptors);
+
+	// TODO: This leaks the memory allocated for the field descriptors
+}
+
+LPCGUID EtwProvider::Guid() const noexcept
+{
+	return m_guid;
+}
+
+EtwProviderEvent* EtwProvider::FindEvent(const char* eventName)
```

Comment:
Return a reference rather than a pointer. We own this event as a provider so we can show that via a reference and avoid the possibility of a null pointer while we're at it. I'm not sure if it makes sense to modify if an event or not, if it doesn't then return a const reference instead.

Reply:
The thinking here is that the method could or couldn't find an event, if it doesn't we need to make an add one. A `std::optional` would be the ideal return here, but we don't have one. Returning a reference means you can't represent the failure state.

Reply:
Ah I see, ok then